corehq/apps/users/audit/change_messages.py

```diff
@@ -1,91 +1,294 @@
+            ]}
         else:
-            message = 'Program: None'
-        return message
+            change_message = {"program": [
+                {"slug": "clear_program"}
+            ]}
+        return change_message
 
     @staticmethod
     def role_change(user_role):
-        message = 'Role: None'
         if user_role:
-            message = f"Role: {user_role.name}[{user_role.get_qualified_id()}]"
-        return message
+            change_message = {"role": [
+                {
+                    "slug": "set_role",
+                    "params": {"id": user_role.get_qualified_id(), "name": user_role.name}
+                }
+            ]}
+        else:
+            change_message = {"role": [
+                {"slug": "clear_role"}
+            ]}
+        return change_message
 
     @staticmethod
     def domain_removal(domain):
-        return f"Removed from domain '{domain}'"
+        return {"domain": [
+            {
+                "slug": "remove_from_domain",
+                "params": {"name": domain}
+            }
+        ]}
 
     @staticmethod
     def registered_devices_reset():
-        return "Registered devices reset"
+        return {"devices": [
+            {
+                "slug": "reset_devices"
+            }
+        ]}
 
     @staticmethod
     def two_factor_disabled_for_days(days):
-        return f"Disabled for {days} days"
+        return {"two_factor": [
+            {
+                "slug": "disable_for_days",
+                "params": {"days": days}
+            }
+        ]}
 
     @staticmethod
     def two_factor_disabled_with_verification(verified_by, verification_mode):
-        return f'Two factor disabled. Verified by: {verified_by}, verification mode: "{verification_mode}"'
+        return {"two_factor": [
+            {
+                "slug": "disable_with_verification",
+                "params": {
+                    "verified_by": verified_by,
+                    "verification_mode": verification_mode
+                }
+            }
+        ]}
 
     @staticmethod
     def password_reset():
-        return "Password reset"
+        return {"password": [
+            {
+                "slug": "reset_password"
+            }
+        ]}
 
     @staticmethod
-    def status_update(action, reason):
-        return f'User {action}. Reason: "{reason}"'
+    def status_update(active, reason):
+        return {"status": [
+            {
+                "slug": "activate_user" if active else "deactivate_user",
+                "params": {
+                    "reason": reason
+                }
+            }
+        ]}
 
     @staticmethod
-    def phone_number_added(phone_number):
-        return f"Added phone number {phone_number}"
+    def phone_numbers_added(phone_numbers):
+        return {"phone_numbers": [
+            {
+                "slug": "add_phone_numbers",
+                "params": {
+                    "phone_numbers": ",".join(phone_numbers)
```

Comment:
why not make this an array and let the formatter handle it?

Reply:
You mean combine the numbers in the raw string. Am i understanding this correctly?
```
In [1]: a = "phone numbers: {', '.join(phone_numbers)}"

In [2]: a.format(phone_numbers=['12222','22222','3333'])
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-2-28f7c91050be> in <module>
----> 1 a.format(phone_numbers=['12222','22222','3333'])

KeyError: "', '"
```

Reply:
ah I see, fine to leave as is

Reply:
That is actually one major thing I wanted to flag about using raw messages like this.
More details in the first point in comment https://github.com/dimagi/commcare-hq/pull/30181#discussion_r684270975

we would have a bunch of methods written to override the default behavior to just format the params differently, which is okay as well but it does makes things closer to just like the previous approach, defeating the point of making thing generic.